src/models/pipeline.go

```diff
@@ -549,83 +550,130 @@ func (m *Pipeline) IDHex() string {
 	return hex.EncodeToString([]byte(m.ID))
 }
 
+type ByPublishTime []*pubsub.ReceivedMessage
+
+func (a ByPublishTime) Len() int      { return len(a) }
+func (a ByPublishTime) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+func (a ByPublishTime) Less(i, j int) bool {
+	return strings.Compare(a[i].Message.PublishTime, a[j].Message.PublishTime) < 0
+}
+
+type ErrorMessages []string
+
+func (em ErrorMessages) Error() error {
+	if len(em) > 0 {
+		return fmt.Errorf(strings.Join(em, "\n"))
+	}
+	return nil
+}
+
 func (m *Pipeline) PullAndUpdateJobStatus(ctx context.Context) error {
 	s := &PubsubSubscriber{MessagePerPull: 10}
 	err := s.setup(ctx)
 	if err != nil {
 		return err
 	}
 
-	jobMap := map[string]*Job{}
+	messagesForJob := map[string][]*pubsub.ReceivedMessage{}
 
 	accessor := m.JobAccessor()
-	err = s.subscribe(ctx, m.ProgressSubscriptionFqn(), func(recvMsg *pubsub.ReceivedMessage) error {
+	subscription := m.ProgressSubscriptionFqn()
+	err = s.subscribe(ctx, subscription, func(recvMsg *pubsub.ReceivedMessage) error {
 		attrs := recvMsg.Message.Attributes
 		jobId := attrs[JobIdKey]
-		job := jobMap[jobId]
-		if job == nil {
-			var err error
-			job, err = accessor.Find(ctx, jobId)
-			if err != nil {
-				return err
-			}
-			jobMap[jobId] = job
+		if messagesForJob[jobId] == nil {
+			messages := []*pubsub.ReceivedMessage{}
+			messagesForJob[jobId] = messages
 		}
+		messagesForJob[jobId] = append(messagesForJob[jobId], recvMsg)
+		return nil
+	})
+	if err != nil {
+		return err
+	}
 
-		if len(recvMsg.Message.Data) > 0 {
-			b, err := base64.StdEncoding.DecodeString(recvMsg.Message.Data)
-			if err != nil {
-				return err
-			}
-			job.Output += "\n\n" + string(b)
-		}
+	for _, recvMsgs := range messagesForJob {
+		sort.Sort(ByPublishTime(recvMsgs))
```

Comment:
FYI: You can use `sort.Slice()` if go 1.8 or later.

ref:
* https://golang.org/pkg/sort/#Slice
* https://golang.org/doc/go1.8#sort_slice
* https://mattn.kaoriya.net/software/lang/go/20161004092237.htm